### PR TITLE
New API for Updating user Language Preferences

### PR DESCRIPTION
**Pull request by sumanvpacewisdom**

### CODE REVIEW

File: src/services/user.js [hunk 1]
```diff
@@ -33,6 +33,7 @@ module.exports = class UserHelper {
 	 * @returns {JSON} - update user response
 	 */
 	static async update(bodyData, id, orgId) {
+		let skipValidation = true
```
Comment (rakeshSgr) [inline src/services/user.js:36]: @sumanvpacewisdom is it internal API ?

Comment (nevil-mathew) [inline src/services/user.js:36]: Can you rename 'skipValidation' to 'skipRequiredValidation'

File: src/generics/utils.js
Comment (nevil-mathew) [inline src/generics/utils.js:1]: Are we not adding the changes needed for session apis ?

File: src/services/user.js [hunk 2]
```diff
@@ -296,4 +296,82 @@ module.exports = class UserHelper {
...
+					responseCode: 'UNAUTHORIZED',
+				})
+			}
+			let defaultOrg = await organizationQueries.findOne(
+				{ code: process.env.DEFAULT_ORGANISATION_CODE },
+				{ attributes: ['id'] }
+			)
+			let defaultOrgId = defaultOrg.id
+			let userModel = await userQueries.getColumns()
+			const filter = {
```
Comment (aks30) [inline src/services/user.js:325]: We should only query the language entity type no?

File: src/services/user.js [hunk 3]
```diff
@@ -296,4 +296,82 @@ module.exports = class UserHelper {
...
+			}
+			const redisUserKey = common.redisUserPrefix + id.toString()
+			if (await utils.redisGet(redisUserKey)) {
+				await utils.redisDel(redisUserKey)
+			}
+			const processDbResponse = utils.processDbResponse(
+				JSON.parse(JSON.stringify(updatedData[0])),
+				validationData
+			)
+			delete processDbResponse.refresh_tokens
```
Comment (aks30) [inline src/services/user.js:365]: Why so much unnecessary code?

File: src/validators/v1/user.js [hunk 4]
```diff
@@ -8,45 +8,50 @@ const filterRequestBody = require('../common')
...
 	},
 
 	share: (req) => {
 		req.checkParams('id').notEmpty().withMessage('id param is empty')
 	},
+
+	updateLanguage: (req) => {
+		req.checkBody('preferred_language')
+			.trim()
+			.isIn(['en', 'hi'])
```
Comment (aks30) [inline src/validators/v1/user.js:54]: This doesn't make sense, tomorrow more lanugages can be added from entity type

**Review by aks30 — CHANGES_REQUESTED** [review]
Reviewed

File: src/validators/v1/user.js (hunk 4 above)
Comment (sumanvpacewisdom) [reply src/validators/v1/user.js:54]: @aks30 this is the existing code and i have just aligned separately

File: src/services/user.js (hunk 3 above)
Comment (sumanvpacewisdom) [reply src/services/user.js:365]: @aks30 
This is the same code for updating the user they have used

File: src/validators/v1/user.js (hunk 4 above)
Comment (aks30) [reply src/validators/v1/user.js:54]: But its not valid.

File: src/services/user.js (hunk 3 above)
Comment (aks30) [reply src/services/user.js:359]: But what if we don't do here.

**Review by aks30 — CHANGES_REQUESTED** [review]
Reviewed on 5 July 7:30 PM

File: src/validators/v1/user.js (hunk 4 above)
Comment (sumanvpacewisdom) [reply src/validators/v1/user.js:54]: OK will remove .
@aks30

File: src/controllers/v1/user.js [hunk 5]
```diff
@@ -61,4 +61,26 @@ module.exports = class User {
 			return error
 		}
 	}
+
+	/**
+	 * Updates preferred language of user
+	 * @method
+	 * @name update
```
Comment (rakeshSgr) [inline src/controllers/v1/user.js:68]: @sumanvpacewisdom change the method name

File: src/controllers/v1/user.js [hunk 6]
```diff
@@ -61,4 +61,26 @@ module.exports = class User {
...
+	 * @param {Object} req -request data.
+	 * @param {Object} req.body - contains user preferred language.
+	 * @returns {JSON} - user updated data.
```
Comment (rakeshSgr) [inline src/controllers/v1/user.js:71]: @sumanvpacewisdom update the return response message

File: src/database/migrations/20240705115500-update_required_column_EntityTypes.js [hunk 7]
```diff
@@ -0,0 +1,29 @@
+'use strict'
+/** @type {import('sequelize-cli').Migration} */
+module.exports = {
+	async up(queryInterface, Sequelize) {
+		// Update the 'required' column for specific existing data to true
+		const valuesToUpdate = ['languages', 'location']
```
Comment (rakeshSgr) [inline src/database/migrations/20240705115500-update_required_column_EntityTypes.js:6]: @sumanvpacewisdom only two columns ?

File: src/services/user.js [hunk 8]
```diff
@@ -296,4 +296,76 @@ module.exports = class UserHelper {
 			throw error
 		}
 	}
+
+	/**
+	 * update user preferred language
+	 * @method
+	 * @name update
```
Comment (rakeshSgr) [inline src/services/user.js:303]: @sumanvpacewisdom update method name

File: src/services/user.js [hunk 9]
```diff
@@ -296,4 +296,76 @@ module.exports = class UserHelper {
...
+	 * @param {Object} bodyData - it contains user preferred language
+	 * @returns {JSON} - updated user response
```
Comment (rakeshSgr) [inline src/services/user.js:305]: @sumanvpacewisdom update the proper response

File: src/services/user.js [hunk 10]
```diff
@@ -296,4 +296,76 @@ module.exports = class UserHelper {
...
+	 */
+	static async setLanguagePreference(bodyData, id, orgId) {
+		let skipRequiredValidation = true
+		bodyData.updated_at = new Date().getTime()
```
Comment (rakeshSgr) [inline src/services/user.js:309]: @sumanvpacewisdom move this line to below when you are going to update, its not require in failure cases

File: src/services/user.js [hunk 11]
```diff
@@ -296,4 +296,76 @@ module.exports = class UserHelper {
...
+				})
+			}
+			let defaultOrg = await organizationQueries.findOne(
+				{ code: process.env.DEFAULT_ORGANISATION_CODE },
+				{ attributes: ['id'] }
+			)
+			let defaultOrgId = defaultOrg.id
+			let userModel = await userQueries.getColumns()
+			const filter = {
+				status: 'ACTIVE',
```
Comment (rakeshSgr) [inline src/services/user.js:326]: @sumanvpacewisdom use common constants

File: src/services/user.js [hunk 12]
```diff
@@ -296,4 +296,76 @@ module.exports = class UserHelper {
...
+				organization_id: { [Op.in]: [orgId, defaultOrgId] },
+				model_names: { [Op.contains]: [userModel] },
+				value: 'preferred_language',
```
Comment (rakeshSgr) [inline src/services/user.js:329]: @sumanvpacewisdom will it be a constant?

File: src/services/user.js [hunk 13]
```diff
@@ -296,4 +296,76 @@ module.exports = class UserHelper {
...
+			}
+			let validationData = await entityTypeQueries.findUserEntityTypesAndEntities(filter)
+			const prunedEntities = removeDefaultOrgEntityTypes(validationData)
+			//validationData = utils.removeParentEntityTypes(JSON.parse(JSON.stringify(validationData)))
+
+			let res = utils.validateInput(bodyData, prunedEntities, userModel, skipRequiredValidation)
```
Comment (rakeshSgr) [inline src/services/user.js:335]: @sumanvpacewisdom use proper naming convention

File: src/services/user.js [hunk 14]
```diff
@@ -296,4 +296,76 @@ module.exports = class UserHelper {
...
+			let userModel = await userQueries.getColumns()
+			const filter = {
+				status: 'ACTIVE',
+				organization_id: { [Op.in]: [orgId, defaultOrgId] },
+				model_names: { [Op.contains]: [userModel] },
+				value: 'preferred_language',
+			}
+			let validationData = await entityTypeQueries.findUserEntityTypesAndEntities(filter)
+			const prunedEntities = removeDefaultOrgEntityTypes(validationData)
+			//validationData = utils.removeParentEntityTypes(JSON.parse(JSON.stringify(validationData)))
```
Comment (rakeshSgr) [inline src/services/user.js:333]: @sumanvpacewisdom remove  commented lines if not required

File: src/services/user.js [hunk 15]
```diff
@@ -296,4 +296,76 @@ module.exports = class UserHelper {
...
+
+			let res = utils.validateInput(bodyData, prunedEntities, userModel, skipRequiredValidation)
+			if (!res.success) {
+				return responses.failureResponse({
+					message: 'SESSION_CREATION_FAILED',
```
Comment (rakeshSgr) [inline src/services/user.js:338]: @sumanvpacewisdom are we creating session here ?

File: src/services/user.js [hunk 16]
```diff
@@ -296,4 +296,76 @@ module.exports = class UserHelper {
...
+			const processDbResponse = utils.processDbResponse(
+				JSON.parse(JSON.stringify(updatedData[0])),
+				validationData
+			)
+			delete processDbResponse.refresh_tokens
+			delete processDbResponse.password
+			return responses.successResponse({
+				statusCode: httpStatusCode.accepted,
+				message: 'PROFILE_UPDATED_SUCCESSFULLY',
+				result: processDbResponse,
```
Comment (rakeshSgr) [inline src/services/user.js:364]: @sumanvpacewisdom change the response message to specific to prefered language settings

File: src/services/user.js [hunk 17]
```diff
@@ -296,4 +296,76 @@ module.exports = class UserHelper {
...
+			)
+			const redisUserKey = common.redisUserPrefix + id.toString()
+			if (await utils.redisGet(redisUserKey)) {
+				await utils.redisDel(redisUserKey)
+			}
+			const processDbResponse = utils.processDbResponse(
+				JSON.parse(JSON.stringify(updatedData[0])),
+				validationData
+			)
+			delete processDbResponse.refresh_tokens
```
Comment (rakeshSgr) [inline src/services/user.js:359]: @sumanvpacewisdom just make the common method to delete un necessary fields before sending to response. 
Create function and use it here, we can make use later

File: src/validators/v1/user.js [hunk 18]
```diff
@@ -8,45 +8,50 @@ const filterRequestBody = require('../common')
...
+
+		req.checkBody('image').optional().isString().withMessage('image field must be string only')
 	},
 
 	share: (req) => {
 		req.checkParams('id').notEmpty().withMessage('id param is empty')
 	},
+
+	setLanguagePreference: (req) => {
+		req.checkBody('preferred_language')
```
Comment (rakeshSgr) [inline src/validators/v1/user.js:52]: @sumanvpacewisdom is it mandatory ? then validate it

File: src/controllers/v1/user.js [hunk 19]
```diff
@@ -61,4 +61,26 @@ module.exports = class User {
...
+	 */
+	async setLanguagePreference(req) {
```
Comment (rakeshSgr) [inline src/controllers/v1/user.js:73]: @sumanvpacewisdom api-doc changes are missing

File: src/database/migrations/20240705115500-update_required_column_EntityTypes.js (hunk 7 above)
Comment (sumanvpacewisdom) [reply src/database/migrations/20240705115500-update_required_column_EntityTypes.js:6]: We have three , and we don't have language preferences to be validated while updating user ,so we are making only 2 entityTypes as true

File: src/services/user.js (hunk 12 above)
Comment (sumanvpacewisdom) [reply src/services/user.js:329]: @rakeshSgr not it cannot be

File: src/validators/v1/user.js (hunk 18 above)
Comment (sumanvpacewisdom) [reply src/validators/v1/user.js:52]: @rakeshSgr yes

File: src/api-doc/api-doc.yaml [hunk 20]
```diff
@@ -20,7 +20,7 @@ paths:
   /user/v1/account/create:
     post:
       summary: Create user
-      tags:
+      x-tags:
```
Comment (rakeshSgr) [inline src/api-doc/api-doc.yaml:23]: @sumanvpacewisdom why its added?

**Pull request merged**